### ISSUE TITLE
feature request: allow functions as arguments in constant literals

### ISSUE THREAD

**Issue by TanklesXL** [gleam-lang/gleam#1632]
_sorry it took me so long to remember to create this issue.  :sweat_smile:_ 

Fundamentally, this feature request is to allow functions to be used as arguments in the creation of compound constant literals such as lists, tuples, and structs.

For example
given types: 

```gleam
type Func(a) =
  fn(a) -> a

type Tup(a) =
  #(Func(a), Func(a))

type Lis(a) =
  List(Func(a))

type Stru(a) {
  Stru(Func(a))
}

```
and the function: 
```gleam
fn int_identity(i: Int) -> Int { i }
```

I would expect to be able to do: 
```gleam
const tup: Tup(Int) = #(int_identity, int_identity)

const lis: Lis(Int) = [int_identity, int_identity]

const stru: Stru(Int) = Stru(int_identity)
```
I would think that **allowing functions as argument when all types are known** should be the base case for this kind of behaviour.

_**note: the rest of this is speculation/thinking out loud**_
Furthermore, with a generic function, for example identity 
```gleam
fn identity(a: a) -> a {
  a
}
```

MIght it be possible to create (a limited form of) constant generics?

```gleam
fn identity(a: a) -> a {
  a
}

const tup: Tup(a) = #(identity, identity)

const lis: Lis(a) = [identity, identity]

const stru: Stru(a) = Stru(identity)
```

**Comment by lpil** [comment]
Great idea, thank you

**Comment by e-hat** [comment]
Would it be ok if I gave a go at adding this? It would be my first contribution to the codebase. I have not fully devised a plan yet for how this would be implemented but I can circle back here with that plan after some investigation.

Also, it might be important to note that, not only are compound literal module constants containing functions currently impossible to create, but non-compound constants that are simply functions are impossible to create as well (although I'm not totally sure why someone would do this). Here is an example of what I mean:

```gleam
// These are from the example from @TanklesXL
// They fail to compile, as we know
const tup: Tup(Int) = #(int_identity, int_identity)

const lis: Lis(Int) = [int_identity, int_identity]

const stru: Stru(Int) = Stru(int_identity)

// This also fails with the same message
const f: Func(Int) = int_identity
```

**Comment by TanklesXL** [comment]
Thanks for the added context @e-hat. I'm not sure what @lpil has in mind with regards to non-compound function literals though so I'm curious where the discussion would lead.

**Comment by TanklesXL** [comment]
@e-hat you make a good point, going forward I'll try not to refer to them as literals unless they actually are 😋

**Comment by TanklesXL** [comment]
I will say as to the original intent of the feature request: 
I only really wanted functions as part of compound constants. 

Having standalone function constants, literal or otherwise, means that there is now more than 1 way to define a function at the top level. This has both pros and cons and probably requires more thought on whether or not it's desirable behaviour.

**Comment by lpil** [comment]
Thank you for your help!

Each time another constant is refer to it would need to immediately expand to the value that that constant holds, that way the problem behaviour in other modules will not arise. This is how it currently works so it should be more straightforward to implement it this way as well.

The main difference will be to change the constant data structure to have a type that refers to a function at the module level.

**Comment by lpil** [comment]
References to other constants and module functions should be permitted at the top level. During type checking we will verify that it is only these that are being referred to.

It would be good to support anonymous functions but I think we should consider that a separate secondary feature rather than part of this ticket.

**Comment by e-hat** [comment]
How's this for a plan: 
- [ ] Add variant `Constant::Name` (in `compiler-core/src/ast/constant.rs`)
- [ ] Parse identifiers into `Constant::Name` (in `compiler-core/src/parse.rs`)
- [ ] Parse function types for module constants (also in `compiler-core/src/parse.rs`)
- [ ] Typecheck/infer types for `Constant::Name` (in `compiler-core/type_/expression.rs`)
- [ ] Update how module constant statements are codegen'd in Elixir and JS to properly output `Constant::Name` definition statements (in `compiler-core/src/{erlang,javascript}.rs`)
- [ ] Update how module constant usages are codegen'd to properly output `Constant::Name` usages (in `compiler-core/src/{erlang,javascript}.rs`)

**Comment by lpil** [comment]
Sounds good! To be consistent with the other syntax trees `Name` should probably be `Var`

https://github.com/gleam-lang/gleam/blob/736473fc00aa015ec5f4a970ba47d8c03a3a755f/compiler-core/src/ast/typed.rs#L41

**Comment by e-hat** [comment]
Ok. I will be working on it on [this branch](https://github.com/e-hat/gleam/tree/1632-function-const-variant).

**Comment by e-hat** [comment]
Oops - should I not mention this issue in git commit messages like that? Hopefully it didn't bother you both if that sent a notification.

**Comment by e-hat** [comment]
Would people be okay with me opening a draft pull request for this? I added type-checking and some other stuff so there are already a good number of changes to review. 

Also, I'm pretty sure the fix for this will also solve #1416.

**Comment by e-hat** [comment]
Haha nvm to that last question, I ended up finishing the feature this morning :-)
